src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -1035,4 +1044,71 @@ protected Object runBlockingSystemCall(Object executable, RubyArray argsArray,
             return foreignToRubyNode.executeConvert(result);
         }
     }
+
+    @Primitive(name = "thread_each_caller_location")
+    public abstract static class EachCallerLocationNode extends PrimitiveArrayArgumentsNode {
+
+        @Child private CallBlockNode yieldNode = CallBlockNode.create();
+
+        @Specialization
+        protected Object eachCallerLocation(VirtualFrame frame) {
+            final RubyProc block = (RubyProc) RubyArguments.getBlock(frame);
+            // Skip the block of `Thread#each_caller_location` + its internal iteration.
+            final int skip = 2;
+            final List<TruffleStackTraceElement> stackTraceElements = new ArrayList<>();
+            final AtomicInteger index = new AtomicInteger(0);
+
+            Truffle.getRuntime().iterateFrames((frameInstance) -> {
+                final RootCallTarget rootCallTarget = (RootCallTarget) frameInstance.getCallTarget();
+                Node location = frameInstance.getCallNode();
+
+                if (location != null) {
```

Comment:
Wondering in what cases `location` might be `null`.

Reply:
I don't know enough about the internal calls to explain that. All I can observe there are empty locations.

Reply:
Checking it from a few examples it seems it's always the first one. Since we need to skip anyways, I simplified the sequence. Let me know if this makes sense.